toolchain/check/context.cpp

```diff
@@ -306,37 +309,146 @@ auto Context::LookupUnqualifiedName(Parse::NodeId node_id,
+
+  auto inst_constant_id = context.constant_values().Get(access_info->base_id);
+  if (!inst_constant_id.is_constant()) {
+    class_type = context.types().TryGetAs<SemIR::ClassType>(
+        context.insts().Get(access_info->base_id).type_id());
+  } else {
+    class_type = context.insts().TryGetAs<SemIR::ClassType>(
+        context.constant_values().GetInstId(inst_constant_id));
+  }
+
+  if (!class_type) {
+    return;
+  }
+
+  // TODO: Support scoped entities other than just classes.
+  auto class_info = context.classes().Get(class_type->class_id);
+  auto is_access_from_instance_method =
+      access_info->highest_allowed_access == SemIR::AccessKind::Private;
+
+  CARBON_CHECK(class_info.is_defined());
+
+  CARBON_DIAGNOSTIC(ClassInvalidMemberAccess, Error,
+                    "Cannot access {0} member `{1}` of type `{2}`",
+                    std::string_view, SemIR::NameId, SemIR::TypeId);
+  CARBON_DIAGNOSTIC(ClassMemberDefinition, Note,
+                    "{0} member `{1}` defined here.", std::string_view,
+                    SemIR::NameId);
+
+  if (access_kind == SemIR::AccessKind::Protected) {
+    if (!is_access_from_instance_method) {
+      context.emitter()
+          .Build(node_id, ClassInvalidMemberAccess, "protected", name_id,
+                 class_info.self_type_id)
+          .Note(member_id, ClassMemberDefinition, "Protected", name_id)
+          .Emit();
+      return;
+    }
+  }
+
+  if (access_kind == SemIR::AccessKind::Private) {
+    if (!is_access_from_instance_method) {
+      context.emitter()
+          .Build(node_id, ClassInvalidMemberAccess, "private", name_id,
+                 class_info.self_type_id)
+          .Note(member_id, ClassMemberDefinition, "Private", name_id)
+          .Emit();
+      return;
+    }
+
+    if (is_parent_access && is_access_from_instance_method) {
+      auto parent_type_id = SemIR::TypeId::Invalid;
+      if (auto base_decl = context.insts().TryGetAsIfValid<SemIR::BaseDecl>(
+              class_info.base_id)) {
+        parent_type_id = base_decl->base_type_id;
+      }
+      if (auto adapt_decl = context.insts().TryGetAsIfValid<SemIR::AdaptDecl>(
+              class_info.adapt_id)) {
+        parent_type_id = adapt_decl->adapted_type_id;
+      }
+
+      context.emitter()
+          .Build(node_id, ClassInvalidMemberAccess, "private", name_id,
+                 parent_type_id)
+          .Note(member_id, ClassMemberDefinition, "Private", name_id)
+          .Emit();
+      return;
+    }
+  }
+}
+
+static auto IsAccessProhibited(std::optional<AccessInfo> access_info,
+                               SemIR::AccessKind access_kind,
+                               bool is_parent_access) -> bool {
+  if (!access_info) {
+    return false;
+  }
+
+  switch (access_kind) {
+    case SemIR::AccessKind::Public:
+      return false;
+    case SemIR::AccessKind::Protected:
+      return access_info->highest_allowed_access !=
+             SemIR::AccessKind::Protected;
```

Comment:
IIUC you're still making changes, but I'll flag != versus < here. Looking at the tests in https://github.com/carbon-language/carbon-lang/pull/4248#discussion_r1744202933, `noninstance_protected_on_self` might catch this.

Reply:
It doesn't seem to be the case?

Reply:
Oh, I see the problem now. Thanks!